src/validation.cpp

```diff
@@ -2677,13 +2677,19 @@ bool CChainState::ActivateBestChain(CValidationState &state, const CChainParams&
     do {
         boost::this_thread::interruption_point();
 
+#ifdef DEBUG_LOCKORDER
+        if (IsRunningInValidationInterfaceQueue()) {
+            throw std::logic_error(
+                "This function cannot be called from within the ValidationInterface "
+                "queue because it may result in a deadlock when attempting to "
+                "drain the same queue below.");
+        }
+#endif
```

Comment:
It occurs to me now that for futureproofing we should add similar code to SyncWithValidationInterfaceQueue(). Will add.